### DISCUSSION TITLE
Microsoft authentication issue using Azure FrontDoor

### DISCUSSION THREAD

**Discussion by guillaumemeyer** [outline/outline#3928] [Self Hosting]
Our Outline server is hosted as a custom container in Azure app service, then exposed through Azure FrontDoor.

Using Microsoft signin, we're blocked by the following error:
- "notice=state-mismatch"
- "Authentication failed – we were unable to sign you in at this time. Please try again."

![image](https://user-images.githubusercontent.com/1385518/183227714-b20e9d9f-b53e-44ba-bb2e-73180d82e500.png)

From the container logs, I can see:
```
2022-08-06T00:37:19.376627810Z {"label":"http","level":"info","message":"  <-- GET /auth/azure.callback?code=REDACTED&state=2f78b4deb4e3a1c6&session_state=6fdfe06b-617a-434e-8597-b29c5bc7ab7a"}
```
The `state` parameter seems present...
But then:
```
2022-08-06T00:37:19.377988124Z {"error":"State not return in OAuth flow","stack":"BadRequestError: State not return in OAuth flow\n    at OAuthStateMismatchError (/opt/outline/build/server/errors.js:113:34)\n ... at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)","level":"error","message":"Error during authentication"}
```

I looked at the following issues without success:
- https://github.com/outline/outline/discussions/2380
- https://github.com/outline/outline/discussions/2089

Any clue?

**Comment by tommoor** [comment]
This means that the server tried to read the `state` cookie from the client to verify the OAuth connection and it was empty/didn't exist. Are you using a proper domain name with SSL?

**Reply by guillaumemeyer** [reply]
The topology looks like this:
1. Client
2. Azure FrontDoor (TLS certificate auto-generated for the public URL app.domain.com)
3. Azure App service (Defined as a target in Azure FrontDoor proxy as service.azurewebsites.net)

Outline responds correctly to the public URL with the login page and the TLS certificate is marked as valid in the browser.

**Reply by guillaumemeyer** [reply]
BTW the error message is quite misleading as it shows the HTTP GET request with the right state value as a query parameter... Now I understand that it's related to cookies.
Is there any reason to control the returned state value from a cookie instead of from the query?
Is that configurable for a specific provider such as Microsoft Azure?

**Reply by guillaumemeyer** [reply]
I'm gonna check if there's any known issue with FrontDoor regarding cookies, but I'm using this service as a proxy and TLS provider for dozens of different services, and it's the first time I'm seeing this error...

**Reply by tommoor** [reply]
> Is there any reason to control the returned state value from a cookie instead of from the query?

The cookie `state` is compared against the query `state` returned from Azure, it's one of the protections in the OAuth process to do so, it's not configurable.
